qip-0008.md

```diff
@@ -0,0 +1,234 @@
+
+$$ S_{uncles} = \sum_{i=0}^N (256 - log(hash_i)) $$
+
+### Computing Security Efficiency Score $f$
+Each node shall record recently uncled entropy and recently accepted entropy
+values for each chain it is paritcipating in. Recentness of this measurement is
+achieved by applying an exponentially weighted moving average to the
+accepted/uncled entropy observed at each new block. This score can be computed
+iteratively with each new block added to the chain. First, the node will
+compute an exponentially weighted moving average of the uncle_entropy and
+accepted_entropy. 
+
+The weighted accepted & uncled entropies must be recorded in each chain, so
+that an aggregate security score can be computed in the prime chain of any
+node, regardless of which subordinate chains a node participates in.
+
+#### Computing Cumulative Accepted & Uncled Entropies in Zone Chains
+In accordance with PoEM consensus, each node measures the entropy reduction
+achieved by the work performed on every block. This entropy is accumulated and
+referred to as $S$. Additionally, in the Quai protocol, dominant chains record
+the cumulative entropy reduction achieved by each subordinate chain, referred
+to as $\Delta S_{sub}$. Each of these values are recorded in the header for
+their parent block, as `ParentS` and `ParentSubDeltaS`.
+
+In order for dominant chains to compute a security score across the network,
+each chain will additionally need to record the uncled and cumulative
+entropies, $S_{uncle}$ & $\Delta S_{sub, uncles}$ respectively. These values
+shall be calculated identically to the accepted entropies, except that the
+computation will be measured on the uncle blocks instead of the accepted
+blocks.
+
+##### Necessary Zone Chain Data In Block Header
+In addition to as $S$ & $\Delta S_{sub}$, which are already recorded in the
+header, each chain shall record $S_{uncle}$ & $\Delta S_{sub, uncles}$. The
+four values recorded in the header (except in the case of zones, which do not
+have subordinate chains to record), are given as follows:
+
+```
+type Header {
+    ...
+    ParentAcceptedS: bigint,
+    ParentUncledS: bigint,
+    ParentAcceptedSubDeltaS: bigint,
+    ParentUncledSubDeltaS: bigint,
+    ...
+}
+```
+
+#### Computing a Network-Wide Security Score
+Before computing a security score, we must exponentially weight the cumulative
+entropies, to prevent the score from becoming overwhelmingly biased by old
+blocks. We can then use these wieghted measurements to compute a security
+score.
+
+##### Computing Exponentially Weighted Entropies
+First compute an error value $\epsilon_i$, as the exponentially weighted uncle
+entropy at block[i]:
+
+$$ \epsilon[i] = S_{uncle}[i-1] * \alpha + S_{uncle}[i]*(1 - \alpha) $$
+
+Next compute the exponentially weighted accepted entropy, $S_{exp}$ at block[i]:
+
+$$ S_{exp}[i] = S[i-1] * \alpha + S[i]*(1 - \alpha) $$
+
+> Note, for efficiency `S` values are recorded and computed in their log2 form
+> (aka bits).
+
```

Comment:
Should clarify that this is only the integer portion of the log2.

Reply:
added